File: majority_element.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        count = 1
        candidate = nums[0]
        for e in nums[1:]:
            if e == candidate:
                count += 1
            elif count > 0:
                count -= 1
            else:
                candidate = e
                count = 1
        return candidate

File: majority_element.java
// Time: O(n) Space: 0(1)
class Solution {
    public int majorityElement(int[] nums) {
        // two ways to go about this
        // Booyer Moore Voting Algorithm or HashMap
        // Lets stick with the theme of arrays and utilize Booyers!

        //instead of starting at 0 lets begin at the first position in the array
        int candidate = nums[0];
        int count = 1;
        for (int e = 1; e < nums.length; e++){
            if (nums[e] == candidate) {count++;}
            else if (count > 0) {count--;}
            else 
                {candidate=nums[e]; count=1;}
        }
        return candidate;
    }
}
